### DISCUSSION TITLE
Using tailwind build as a library

### DISCUSSION THREAD

**Discussion by parris** [tailwindlabs/tailwindcss#11282] [Help] [answered]
Hi all,

I've been working on a no-code editor that can use tailwind. The final piece of the puzzle for me right now is making a build whenever a user publishes a page. I've been trying to read through tailwind's source a bit to figure out how to make this happen.

My current thought is to extract all the css classes myself out of the documents that get generated via my wyswyg editor, and creating a config with all those items on the `safelist`. After that I'd hop to get a string back representing the compiled css, which I'd upload to GCS/S3.

Any tips/ideas?

Thanks!

**Comment by wongjn** [comment, marked as answer]
You could try using the PostCSS API to programmatically run Tailwind:

```js
const postcss = require('postcss');
const tailwindcss require('tailwindcss');
const baseConfig = require('./tailwind.config');

const sourceCSS = '@tailwind base; @tailwind components; @tailwind utilities';
const config = {
  presets: [baseConfig],
  safelist: mySafelist,
};
postcss([tailwindcss(config)])
  .process(sourceCSS)
  .then((css) => {
    // `css` is a string of the compiled CSS.
  });
```

**Reply by parris** [reply]
oh wow - super close now. At this point I know I'm asking a postcss question, but the output produced is:

`@tailwindcss base; @tailwindcss components; @container main (min-width: 1024px){ .\\@lg\\/main\\:underline{ text-decoration-line: underline }`

This isn't valid css with all the `@` symbols. Should I be doing something else to further process the result.

**Reply by wongjn** [reply]
No, the `@tailwindcss base; @tailwindcss components;` are invalid – you may have meant `@tailwind base; @tailwind components;`. I've updated the code.

**Reply by parris** [reply]
Awesome! That worked - thanks so much

**Comment by AdsonCicilioti** [comment]
@parris I'm also working on a visual editor. How nice to know that!

@wongjn I know I can open another topic but I'm trying to imagine how to create a library of reactjs components with properties that the developer chooses to define a certain style variant of a button, for example.

My question is how to deliver to the developer a way to identify based on the components and properties he used in the app and thus compile only the css used?

Could you show me a way?

**Reply by wongjn** [reply]
The consumer could consider adding paths to `content` of the library files they use.

**Reply by AdsonCicilioti** [reply]
Ok, I'm going to create some function that checks, in the app where I installed my ui lib, which components were used and include them in the `content`.

**Reply by vaynevayne** [reply]
I had the same idea, but ran into some problems.
1. how to provide self theme color and dark mode ? Need to support runtime changes
2. Card component has a `.card-title`, how to allow user to modify its style, mantine-ui's way is `classNames=`, how to complete same feature ?
